test-lib-openshift.sh

```diff
@@ -291,4 +291,36 @@ function ct_os_cluster_running() {
   oc status &>/dev/null
 }
 
+# ct_os_test_s2i_app IMAGE APP CONTEXT_DIR EXPECTED_OUTPUT [ENV_PARAMS, ... ]
+# --------------------
+# Runs [image] and [app] in the openshift and optionally specifies env_params
+# as environment variables to the image. Then check the http response.
+# Arguments: image - prefix or whole ID of the pod to run the cmd in
+# Arguments: app - url or local path to git repo with the application sources.
+# Arguments: context_dir - sub-directory inside the repository with the application sources.
+# Arguments: expected_output - PCRE regular expression that must match the response body.
+# Arguments: env_params - environment variables parameters for the images.
+function ct_os_test_s2i_app() {
+  local image_name=${1}
+  local app=${2}
+  local context_dir=${3}
+  local expected_output=${4}
+  shift 4
+  local image_name_no_namespace=${image_name##*/}
+  local service_name="${image_name_no_namespace}-testing"
+
+  ct_os_new_project
+  # Create a specific imagestream tag for the image so that oc cannot use anything else
+  ct_os_upload_image "${image_name}" "${service_name}"
+
+  ct_os_deploy_s2i_image "${service_name}" "${app}" \
+                          --context-dir="${context_dir}" \
+                          --name "${service_name}" $@
+
+  ct_os_wait_pod_ready "${service_name}" 300
+  local ip=$(ct_os_get_service_ip "${service_name}")
+  ct_test_response "http://${ip}:8080" 200 "${expected_output}"
```

Comment:
Could we pass a custom response check function instead? Or in addition to this?

Reply:
imo it would make more sense to override this since some images do not provide any content available via http so the failure would have to be ignored.

Reply:
At this point I only added few more things to be configurable (port, protocol, ...). Would that address the use case you have in your mind?
For the other use cases, my thinking was more like to let users write own similar function. This one was meant for images that we might check using `curl` request.

Reply:
My idea was to use this for all of our s2i enabled images, so for example to be able to run `pg_isready` instead of `ct_test_response`.

In any case if this would be too much trouble to accomplish I am ok with providing a component specific test case for s2i builds.

Reply:
@pkubatrh there is now `ct_os_test_s2i_app_func()` function that accepts a check function as argument. Your case with `pg_isready` might then be implemented like this:
```
test_s2i_pg() {
  ct_os_test_s2i_app_func "${IMAGE_NAME}" pg_test . "pg_isready -h <IP>" -e "POSTGRESQL_ADMIN_USER=standa"
}
```
What do you think?

Reply:
That is what I had in mind, thanks!
Although I am not sure the `ct_os_test_s2i_app` function is necessary. I feel like it would make more sense to use the `ct_os_test_s2i_app_func` instead (with the ct_test_response as the function argument).

Reply:
One reasoning for this is that `ct_os_test_s2i_app` requires to source `test-lib.sh` to work. But as nothing inside the actual test script might directly use the library it might be not immediately clear what the library is sourced for.